weather_tools.py

```diff
@@ -0,0 +1,131 @@
+        stations_in_state[states[i]] = list(df_stations_US[df_stations_US.ST == states[i]].stationID.values)
+    return stations_in_state
+stations_in_state = generate_stationID_dict()
+
+
+
+def df_from_file(stationID,YEAR):
+    ''' This takes the stationID and year and outputs a dataframe with the Temperature readings.
+    Note that some tempertaure readings are missing in general.'''
+    
+    url_common = "ftp://ftp.ncdc.noaa.gov/pub/data/noaa/"
+    url_ftp = url_common + str(YEAR)+"/"+ stationID + "-"+  str(YEAR)+".gz"
+    
+    WBAN = re.findall(r'(.*)-(.*)',stationID)
+    WBAN = WBAN[0][1]
+    
+    #df = df_tmp[0].str.extractall(r'24 HR AVG TEMP \(F\):(\s)?(?P<Temperature>\d\d)')
+
+    def extract_Temp(row):
+        result = re.findall(r'24 HR AVG TEMP \(F\):(\s)?(?P<Temperature>\d\d)',row)
+        if result == []: 
+            result = [(np.nan,np.nan)]
+        return result[-1][1]
+
+    def extract_date(row):
+        date = [np.nan,np.nan,np.nan]
+        date = re.findall(WBAN + r'(\d\d\d\d)(\d\d)(\d\d)',row)
+        if date != []: 
+            date =  date[0][0]+ date[0][1] +  date[0][2]
+        return date
+    try:
+        df = pd.read_csv(url_ftp, compression='gzip', header=None)
+        df['TEMPERATURE'+ stationID] = df[0].apply(extract_Temp)
+        df['DATE'] = df[0].apply(extract_date)
+        df.dropna(inplace =True)
+        df.drop(columns = [0], inplace=True)
+        df['DATE'] = pd.to_datetime(df['DATE'])
+        df['TEMPERATURE'+ stationID] = df['TEMPERATURE'+ stationID].astype('int64')
+        df.reset_index(inplace =True)
+        df.drop(columns = ['index'], inplace=True)
+        return df
+    except:
+        #print('No good')
+        DATE = pd.date_range(str(YEAR)+'-01-01', str(YEAR)+'-12-31')
+        df = pd.DataFrame(data =DATE, columns =['DATE'])
+        return df
+
+
+def all_stations_for_state_year(state,year):
+    '''This function reads data from all the stations in a particular state for a particular year 
+    it returns a dataframe with all the dates in that year temperatures. last column contains mean temperatures. '''
+    
+    df_list = []
+    DATE = pd.date_range(str(year)+'-01-01', str(year)+'-12-31')
+    df = pd.DataFrame(data =DATE, columns =['DATE'])
+    for stID in stations_in_state[state]:
+        #print(stID)
+        xx = df_from_file(stID,year)
+        # put code here which gets rid of leap years?
+        df_list.append(xx)
+    for frame in df_list:
+        df =df.merge(frame, on = 'DATE', how ='left')
+    df.set_index('DATE', inplace = True)
+    df['MEAN_T'] = df.mean(axis =1)
+    df['M/D'] = df.index.strftime('%m/%d')
+    return df
+
+
+
+def state_years(state, start_year, end_year):
+    '''Returns a single df with the mean temp data for all years between start-year and end year'''
+    years  = range(start_year, end_year+1,1)
+    df_list = []
+    for year in years:
+        xx=all_stations_for_state_year(state,year)
```

Comment:
Let's try to be consistent with PEP 8 guidelines in terms of spacing
e.g., x = y, not x=y. You have a mixture of styles.